### ISSUE TITLE
Integrate with TCP data via eBPF

### ISSUE THREAD

**Issue by ThatcherT** [FoxIO-LLC/ja4-nginx-module#8]

**Comment by vlvkobal** [comment]
We should key the map by `4-tuple + l4proto` so TCP and UDP flows don’t collide (the same 4-tuple can exist for both). Within a single protocol, `l4proto` won’t prevent collisions from fast tuple reuse (high CPS / NAT can reuse the same tuple quickly), so we should also add a per-connection discriminator - ideally a socket cookie, otherwise a short TTL + generation/epoch.

**Comment by lynch1981** [comment]
I found another approach: TCP_SAVE_SYN / TCP_SAVED_SYN.
Enable on the listen socket, getsockopt on the accepted fd in the nginx worker. Kernel returns the original IP+TCP SYN headers; compute JA4T in userspace.
Pros: no eBPF/maps; stock kernel ≥ 4.3; cheap; same path as the accepted connection.
Cons: empty under SYN cookies (backlog overflow); one-shot read in the accepting worker; needs a small nginx change; no fingerprint for SYNs that never complete.
@vlvkobal

**Comment by lynch1981** [comment]
I have set up a workflow to capture SYN-ACK packets for JA4TS. This makes total sense when Nginx acts as a proxy, as JA4TS on a plain origin server doesn't provide much value.

```md
# eBPF Architecture: Capturing TCP SYN-ACK and Syncing to User Space (Nginx)

This document outlines the architectural design for leveraging eBPF technology to intercept raw TCP handshake 
packets (SYN-ACK) at the underlying network interface (XDP/TC) and securely pass them up to a user-space application 
like Nginx.

---

## 1. Core Pain Points & Challenges

1. **Unassociated Wire Data**: Packets captured at the network ingress (XDP/TC ingress) are raw wire packets without an 
   attached kernel socket reference yet (`sk_buff->sk` is not linked).
3. **NAT Interference**: Network Address Translation may alter IP addresses and ports, 
   breaking traditional 5-tuple tracking.
---

## 2. Solution: "Dual Bridges" and Lifecycle Handover

To bridge this gap, we employ a data migration scheme that combines **`ack_seq`** (the kernel-internal alignment anchor)
with the **`cookie`** (the user-space identification credential).

### Core Map Design
* **Temporary Capture Map (`ack_seq -> raw_syn_ack`)**: Used at the network layer to precisely intercept and store the 
  raw packet.
* **Final Consumption Map (`cookie -> raw_syn_ack`)**: Used by user-space Nginx for direct lookup and retrieval.

---

## 3. Complete Execution Workflow

### Stage 1: Client Connect Phase (Pre-allocation & Placeholder)
* When Nginx triggers `connect()`, the `sockops` hook is invoked.
* The kernel calculates the expected `ack_seq` that the peer will echo back (typically `local_ISN + 1`).
* **Index Setup**: An association placeholder is created in the kernel map in anticipation of the incoming packet.

### Stage 2: XDP / TC Hook Phase (Precise Interception & Filtering)
* When the incoming SYN-ACK packet arrives at the network card, the eBPF program parses the **`ack_seq`** directly
  from the TCP header.
* **Efficient Filtering**: It looks up the temporary capture map using this `ack_seq`:
  * **Hit**: It confirms this is the target connection's handshake packet and writes the `raw syn-ack` data into 
     the map.
  * **Miss**: It treats the packet as unrelated traffic.

### Stage 3: Connection Establishment & Data Migration (Smooth Handover)
* Once the connection is successfully established (triggering the `BPF_SOCK_OPS_ACTIVE_ESTABLISHED_CB` callback),
  the eBPF program **simultaneously possesses** the socket's **`cookie`** and its corresponding **`ack_seq`**.
* **Seamless Relocation**:
  1. The program migrates the `raw syn-ack` data from the temporary map into the final consumption map keyed by 
     the `cookie`.
  3. It cleans up and deletes the obsolete `ack_seq` temporary table entry to free up memory.
### Stage 4: User-Space Retrieval (Nginx)
* After the connection completes, Nginx calls getsockopt to fetch its socket cookie:

  `getsockopt(fd, SOL_SOCKET, SO_COOKIE, &cookie, &len);`

* Armed with this cookie, Nginx directly queries the final shared eBPF map to retrieve the raw SYN-ACK packet captured
  at the lowest level.
```

**Comment by vlvkobal** [comment]
Looks good to me.

One possible simplification: in the initial `sockops` callback, we may already have both the expected ACK and the socket cookie, so TC can store the SYN-ACK directly by cookie and avoid the later migration.

I’d also start with TC ingress only. It should work well for both physical and virtual interfaces, while XDP has known performance drawbacks on `veth`. We can add XDP later if there’s a real need for it.

**Comment by lynch1981** [comment]
Two practical challenges with this setup:

- Nginx requires root privileges to inject eBPF bytecode into the kernel.
- Users must manually specify the network interface.

The first issue is manageable, but the second adds too much configuration friction for users.

**Comment by vlvkobal** [comment]
We can probably avoid requiring users to specify the interface in the common case.

For TC, instead of trying to guess the "right" interface, we could enumerate interfaces in nginx's network namespace and attach to all relevant non-loopback ones, with an explicit interface setting only as an override. We may see the same packet on multiple interfaces, but we can just deduplicate them.

**Comment by lynch1981** [comment]
I have a few concerns regarding interface enum and hook points:

- Dynamic Interfaces: Enum them at startup won't cover virtual interfaces created dynamically at runtime
- Hook point: Attaching too early(eg., physical) means that packet processing occurs before early fast-drop (such as ACK flood mitigation), which adds unnecessary overhead.

**Alternative (BPF_PROG_TYPE_NETFILTER)**: Attaching via BPF_PROG_TYPE_NETFILTER avoids interface enumeration entirely and sits late enough to benefit from early drops before reaching the packet mangling stage. While it requires Linux 6.4+, this is already standard across most current mainstream distributions.

**Comment by vlvkobal** [comment]
Yeah, I think using Netfilter allows us to make the implementation simpler. While we can subscribe to rtnetlink and handle interfaces appearing or disappearing dynamically, `BPF_PROG_TYPE_NETFILTER` avoids interface management entirely.

Good point about the hook placement as well. Compared to attaching TC ingress to all relevant interfaces, running later in the packet path should avoid some unnecessary processing for traffic that gets filtered out earlier.

Since we’re not trying to maximize compatibility with older kernels at this stage, I think the Linux 6.4+ requirement is acceptable.

@igr001-galactica, do you see any concern with requiring Linux 6.4+ for JA4TS at this point? This would exclude some current OS releases, but we can revisit older-kernel support later if there’s actual demand.
